python/pyarrow/scalar.pxi

```diff
@@ -371,14 +371,29 @@ cdef class FixedSizeBinaryValue(ArrayValue):
 
 cdef class StructValue(ArrayValue):
 
+    cdef void _set_array(self, const shared_ptr[CArray]& sp_array):
+        self.sp_array = sp_array
+        self.ap = <CStructArray*> sp_array.get()
+
+    def __getitem__(self, key):
+        cdef:
+            CStructType* type
+            int64_t index
+
+        type = <CStructType*> self.type.type
+        index = type.GetChildIndex(tobytes(key))
+
+        if index < 0:
+            raise KeyError(key)
+
+        return pyarrow_wrap_array(self.ap.field(index))[self.index]
```

Comment:
You are passing an int64_t to a function that takes a (32-bit) int, hence the warning about truncation which is turned into an error.